### PR TITLE
Add PROG-INITIAL-REWIRING-HEURISTIC

### LINKED ISSUE #361: Stochastic failure to compile 'disconnected' program on connected topology.
This came up while running quantum volume, e.g. 
```python
qc = get_qc('Aspen-4-8Q-A', as_qvm=True, noisy=False)
measure_quantum_volume(qc, num_circuits=200, depths=[4])
```
I've been able to isolate the problem to the point where for `Program` prog = 

```
DEFGATE LYR0_RAND0:
    0.3453183205304793-0.12696801053233828i, 0.5265386376910978+0.1931838984531052i, -0.29052305165907266-0.594441852976343i, -0.1506008541194645-0.29937609929246506i
    -0.1593953114488653-0.20866235427010307i, -0.16168835357073963+0.2520089033062991i, 0.239834418759946-0.2438490569066952i, -0.7460675648562454+0.4096362084408356i
    -0.12856619973598493+0.7223359877459085i, 0.4807485314801305-0.24994946297635628i, 0.20519011099777268+0.10677911378343408i, -0.33834032003758086-0.011341762443828618i
    -0.04332809717096084+0.5054816193435523i, -0.5205466694277492+0.1766995012896463i, -0.61250636255294-0.12877703191585974i, -0.18249325797051483-0.12396815063332628i

DEFGATE LYR0_RAND1:
    -0.21163731076953685+0.1556329927465134i, -0.3001566064245327-0.33802345390852884i, 0.7431016288284251-0.37714645363637356i, -0.1772723072697441+0.027734947336669865i
    0.20484396673693783+0.5578246542068888i, 0.5182948848663171+0.13152502746494504i, 0.23098272365238026-0.16896663770943743i, 0.3885456697969944-0.3578710705261025i
    0.05563181510118506-0.18850346757941397i, -0.11536779759544692-0.5264188304392133i, -0.09087636334777575-0.0776098534213045i, 0.7836539637683549+0.20627576785173588i
    -0.6294273438803477-0.378221347023483i, 0.4673940664514512-0.028879206744236774i, -0.14087341305332418-0.4353511883395492i, -0.009238897502739374-0.17893372117702133i

DEFGATE LYR1_RAND0:
    -0.3161540949620567+0.33398130454144603i, 0.5730670791443696-0.1988884256830049i, -0.5087925775965276+0.27789976292939966i, -0.24656615939351895+0.153777622475405i
    -0.1632214729433775-0.0010423436021771426i, -0.5129429781831205-0.01795777167161915i, -0.008356857141457522+0.19784529019496025i, -0.8090085880478683-0.12734674087079353i
    -0.11265894185061123-0.5799816845391261i, 0.1548977424550084-0.54638140790208i, 0.12422194053734126-0.32309325340107314i, -0.20698652962325306+0.40711115257618546i
    -0.167879038799936+0.6201665723798677i, 0.09311177060297278-0.1934134978639262i, 0.4118499852741229-0.5790064659037981i, -0.1488227834117348-0.1188038259582522i

DEFGATE LYR1_RAND1:
    0.4331078161124846+0.4400543862509129i, 0.4208000051589553+0.07896457524190649i, 0.04659789916360604+0.1708617876594143i, -0.5300829080332233+0.35086849987011387i
    -0.42449619979965747+0.08671944279964774i, 0.6581078858490882-0.2647263174878278i, 0.39128728826745485+0.2806321921007687i, 0.23781915262335368-0.1438004987865976i
    -0.4232127234300293-0.18918482804981168i, -0.3759713551069016-0.29388390287934874i, 0.08269952732435294+0.4720324183994802i, -0.4609825680720156+0.3394394488331883i
    0.3394399700609284-0.31769965320647037i, -0.1408431691348342+0.2568013481335408i, 0.6850164435701483+0.19461082268763671i, -0.19563462718700628-0.39072995631142066i

DEFGATE LYR2_RAND0:
    -0.22218188617649237-0.0845177683097678i, -0.4378181601246266-0.40116170852379107i, 0.07804446139150856-0.6442557991967369i, -0.2577262698350556+0.32139881091681616i
    0.5541388296445591-0.32361507550473234i, -0.010687152218892604+0.035527090940233805i, -0.44114310527569184+0.06660655707816715i, -0.6184958459233343+0.0724311174184941i
    0.4696666887659151-0.5379348381961861i, -0.2724633404450848-0.06562754898725322i, 0.11724356313356438-0.17183768191187238i, 0.580286744330692-0.17745173027874905i
    0.008676930289401058-0.14717139287400408i, 0.7298800494823195+0.18638674620598022i, -0.1452494972342694-0.5616299457841333i, 0.15244564675876854+0.22590949192700402i

DEFGATE LYR2_RAND1:
    0.3824355696777235-0.3757101068899931i, -0.013366724791673529-0.41999068627936476i, -0.36327557600486704-0.4408993015214449i, 0.44172932405430937-0.12053203225563681i
    0.5394197315390511+0.2236580141397367i, -0.020634927703455513-0.4264815218878451i, 0.19944007298892202+0.5306429469127093i, -0.09369437137945788-0.38282403547071775i
    -0.356313063910504-0.2233618076875183i, 0.30011163175716876-0.6890860900734945i, -0.008432553745884905+0.162308113452546i, -0.26235534515240944+0.4037307438371577i
    0.4306886563012079+0.09615706552368156i, -0.24477619281644133+0.12765128176165674i, -0.5467994036165233+0.16393828876627842i, -0.31233256056459546+0.5528416473732699i

DEFGATE LYR3_RAND0:
    -0.6107362398210048-0.17616940102131937i, 0.4056086723035124+0.17137280242966307i, -0.06655060982733223+0.513904829547733i, 0.18513549612928504+0.3150813317617598i
    0.3453971867957289+0.09199387685625149i, 0.3428004431741377-0.5336756926884675i, -0.3775159368462484-0.07790591415469098i, 0.10125625998622098+0.5577414965269092i
    0.11373419179727041+0.6407388487190518i, -0.3848758125803704+0.2561499109204719i, -0.2863315034815874+0.46524446900253735i, -0.1506531472112193+0.2040626785973242i
    -0.20779766334764202-0.03926690659204642i, -0.4334687741083179+0.04640832502634282i, 0.43427876688484784-0.3095949535690841i, 0.019760785058551254+0.6931029931594835i

DEFGATE LYR3_RAND1:
    -0.00971498061356324+0.09768179420502382i, -0.06646397232008017-0.05715256500177514i, 0.8112484223853667-0.47275154246627493i, -0.3166682979326618-0.027985286976583836i
    0.560149666182814+0.5834027465471086i, -0.19973095526135018-0.02508027116974949i, 0.05425991501275074-0.06761895733683881i, 0.4732018291532875-0.2718743816958181i
    0.16221703189971087+0.043870469379573444i, 0.46707133721150357-0.8360021093479995i, -0.1190423325835249-0.05254644090175579i, -0.154717061906105-0.11762772188558064i
    -0.24570211782816695-0.4976233101472996i, 0.0015842635478745986-0.18637859167768692i, 0.17069305139289953-0.25455292438373683i, 0.7225858004918309-0.2029740697416773i

LYR0_RAND0 1 10
LYR0_RAND1 10 11
LYR1_RAND0 1 2
LYR1_RAND1 2 10
LYR2_RAND0 2 1
LYR2_RAND1 1 10
LYR3_RAND0 2 1
LYR3_RAND1 1 10
DECLARE ro BIT[4]
MEASURE 1 ro[0]
MEASURE 2 ro[1]
MEASURE 10 ro[2]
MEASURE 11 ro[3]
```
calling 

```python
native_quil = qc.compiler.quil_to_native_quil(prog)
```
sometimes causes the error 

```
error: Unhandled error in host program:
Failed to select a SWAP instruction. Perhaps the qubit graph is disconnected?
```
Specifically, running the loop
```python
for _ in range(100):
    native_quil = qc.compiler.quil_to_native_quil(prog)
```
usually fails at a random iteration < 100. 

~If important I'll sketch the topology of 'Aspen-4-8Q-A'. Note that qubits 1,2 are disconnected from 10,11.~

```
1--2--15--16--17--10--11
          |
         14
```
The problem remains for, e.g. `qc = get_qc('Aspen-4-16Q-A', as_qvm=True, noisy=False)`

### PR DESCRIPTION
**Pull request by appleby**
Per ecpeterson's comments in #361, update `COMPILER-HOOK` to call a new function, `PROG-INITIAL-REWIRING-HEURISTIC`, to try and determine whether a `:NAIVE` initial rewiring strategy is likely to succeed, and otherwise fallback to `*INITIAL-REWIRING-DEFAULT-TYPE*`, which now defaults to `:PARTIAL`, rather than `:NAIVE`.

This commit also adds a bunch of Aspen-*.qpu files exported from `pyquil.api._devices.get_lattice`. The idea is that these QPU files can be used in testing as representative of real Aspen chipspecs.

Additionally, add a (very basic) chip-reader test and the stub of a regression test for #378.

Fixes #361

~~Note that the failing test in `QUILC-TESTS` is due to #375; therefore, this PR should not be merged until after #375.~~
Rebased onto the changes in #375, so tests are happy now.

### CODE REVIEW

**Comment by appleby** [comment]
I also tested these changes against the Python/pyquil example programs provided by kylegulshen and msohaibalam in #361, all three of which now report success rate of 100% if no `INITIAL_REWIRING` pragma is provided.

File: src/addresser/initial-rewiring.lisp [hunk 1]
```diff
@@ -154,17 +154,37 @@ If no
...
-  ;; If we didn't find anything, no instructions, nothing!, then just
-  ;; return the default.
-  *initial-rewiring-default-type*)
+       (return-from prog-rewiring-pragma nil)))))
+
+(defun %naively-applicable (gate-app chip-spec
+                            &aux (qubit-indices (application-qubit-indices gate-app)))
+  "Return true if the given GATE-APP's arguments correspond to a valid HARDWARE-OBJECT in CHIP-SPEC."
+  (and (notany (a:curry #'chip-spec-qubit-dead? chip-spec) qubit-indices)
+       (lookup-hardware-object-by-qubits chip-spec qubit-indices)))
```
Comment (appleby) [inline src/addresser/initial-rewiring.lisp]: Let me know if this test needs to change.

File: tests/chip-specification-tests.lisp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  (quil::read-chip-spec-file (merge-pathnames file-name *qpu-test-file-directory*)))
+
+(deftest test-read-chipspec ()
+  "Test that READ-CHIPSPEC can parse some representative Aspen-* qpu files."
+  (finish-output)
+  (let ((test-files (uiop:directory-files *qpu-test-file-directory* #P"*.qpu")))
+    (is (not (null test-files)))
+    (dolist (file test-files)
+      (format t "~&    Testing file ~A~%" (pathname-name file))
+      (not-signals error (quil::read-chip-spec-file file)))))
```
Comment (appleby) [inline tests/chip-specification-tests.lisp]: This test could definitely be improved, but this was something of a side-quest from the main thrust of this PR, and at least we exercise the code path and now have a place to add more/better tests on some bright, hypothetical, future day.

File: src/addresser/initial-rewiring.lisp (hunk 1 above)
Comment (ecpeterson) [reply src/addresser/initial-rewiring.lisp]: I think you should sort of invert these checks: hardware objects of all orders can be marked as dead, so I'd prefer to check that there is a hardware object there, and then that it itself (and not the 0th order hardware objects embodying its qubits!) is still alive.

**Comment by appleby** [comment]
> "32 files changed" had me briefly spooked!

Heh. I could always pair down the set of Aspen chip files if we feel like checking them all in is overkill. I opted to keep them "just in case"

File: src/addresser/initial-rewiring.lisp (hunk 1 above)
Comment (appleby) [reply src/addresser/initial-rewiring.lisp]: I'm struggling to remember what my rationale was here, but I vaguely recall `lookup-hardware-object-by-qubits` returning nil for dead hardware objects? I do recall that I originally did not have the `chip-spec-qubit-dead?` check at all and was expecting the hardware object to be "dead" if any of qubits were dead, but some of the tests failed and realized I need to check both?

Like I said, my memory is fuzzy so I'll double check this...

Comment (appleby) [reply src/addresser/initial-rewiring.lisp]: Ok, refreshed my memory. Basically the chip-reader will skip any hardware objects of order > 0 that are either themselves dead, or for which any of their associated qubits are dead, which means `LOOKUP-HARDWARE-OBJECT-BY-QUBITS` will return `NIL`. This is not the case for order zero objects, however. In fact, if any qubits are missing from the "1Q" layer, the chip-reader will fill them in with "dead" qubits. Hence why I thought that checking `CHIP-SPEC-QUBIT-DEAD?` was required here.

But perhaps I was too narrowly focused on chip-specs produced by the chip-reader and maybe those constraints don't always apply. I was also influenced by the fact that the `CHIP-SPEC-QUBIT-DEAD?` predicate existed, but there doesn't appear to be a corresponding `HARDWARE-OBJECT-DEAD?` predicate, which made me think that maybe only order 0 objects could be both present and dead...

Anyway, I see your point now. I'll just add a `HARDWARE-OBJECT-DEAD?` predicate and check that directly here.

Comment (appleby) [reply src/addresser/initial-rewiring.lisp]: Done. Note that I dropped the `(notany (a:curry #'chip-spec-qubit-dead? chip-spec) qubit-indices)` altogether. Do we want to retain that to guard against cases where the chipspec has an "alive" order N>0 object but one or more of it's qubits are dead?  Or is that sufficiently verboten that it's not worth checking?

**Comment by ecpeterson** [comment]
I might flip-flop on this advice: I think chip-reader is doing the right
thing with just omitting the dead higher-order object. If we really
supported this, we’d have to pepper the addresser (at least—perhaps the
whole codebase) with `hardware-object-dead?`s.

At any rate, it is verboten to be a live object with dead subobjects—so if
we’re hewing to what the reader is doing, you could still rewrite the
predicate as “missing or (order 0 and dead)”.

**Comment by appleby** [comment]
> I might flip-flop on this advice: I think chip-reader is doing the right thing with just omitting the dead higher-order object. If we really supported this, we’d have to pepper the addresser (at least—perhaps the whole codebase) with `hardware-object-dead?`s. At any rate, it is verboten to be a live object with dead subobjects—so if we’re hewing to what the reader is doing, you could still rewrite the predicate as “missing or (order 0 and dead)”.

Sounds reasonable. I switch back to that then, but only check for dead order-0 objects.

File: src/addresser/initial-rewiring.lisp [hunk 3]
```diff
@@ -154,17 +154,38 @@ If no
...
-       (return))))
-  ;; If we didn't find anything, no instructions, nothing!, then just
-  ;; return the default.
-  *initial-rewiring-default-type*)
+       (return-from prog-rewiring-pragma nil)))))
+
+(defun %naively-applicable-p (gate-app chip-spec
+                              &aux (qubit-indices (application-qubit-indices gate-app)))
+  "Return true if the given GATE-APP's arguments correspond to a valid HARDWARE-OBJECT in CHIP-SPEC."
+  (and (not (null (lookup-hardware-object-by-qubits chip-spec qubit-indices)))
```
Comment (ecpeterson) [inline src/addresser/initial-rewiring.lisp]: 👍

File: tests/initial-rewiring-tests.lisp [hunk 4]
```diff
@@ -82,3 +82,70 @@ HALT
...
+                 "RZ(-0.38636880263183865) 2"
+                 "CNOT 1 2"
+                 "CNOT 1 7"
+                 "RZ(-0.23607918922895188) 7"
+                 "CNOT 1 7"
+                 "CNOT 1 14"))
+        ;; TODO:(appleby) switch the un-altered Aspen-4-10Q-A.qpu file once this issue is resolved.
+        ;; We use a slightly doctored version of the Aspen-4-10Q-A.qpu chipspec file to work around
+        ;; gh-378. Specifically, we altered the original to make sure none of the specs contain
+        ;; f1QRB values > 1.0. See https://github.com/rigetti/quilc/issues/378
```
Comment (ecpeterson) [inline tests/initial-rewiring-tests.lisp]: 👍better mention #378 in a GH comment so that we don't forget about the effect here

**Comment by ecpeterson** [comment]
I feel good about merging this if you're all set.

File: tests/initial-rewiring-tests.lisp (hunk 4 above)
Comment (appleby) [reply tests/initial-rewiring-tests.lisp]: I already left myself a reminder comment over there :)

**Comment by appleby** [comment]
> I feel good about merging this if you're all set.

Yep, good to go.

**Pull request merged**